elastic/apm-server
# datastream: Use internal metrics on empty service.name

## Motivation/summary

Checks that `Service.Name` is not empty before using the field's value
in the datastream `metrics-apm.app.<Service.Name>`.

## Checklist



~- [ ] Documentation has been updated~

## How to test these changes

1. `docker-compose up -d`
2. Install APM Integration through Kibana.
3. Create [an APM Agent Configuration](http://localhost:5601/app/apm/settings/agent-configuration/create) setting the max_transactions to `100` for all agents.
4. `make`
5. `./apm-server -E output.elasticsearch.username=admin -E output.elasticsearch.password=changeme -e`
6. Run this Go program
```go
package main

import (
	"context"
	"fmt"
	"os"
	"time"

	"go.elastic.co/apm"
)

func main() {
	version := "undefined"
	if len(os.Args) > 1 {
		version = os.Args[1]
	}
	name := fmt.Sprintf("apm-server-%s", version)
	tx := apm.DefaultTracer.StartTransaction(name, "type")
	ctx := apm.ContextWithTransaction(context.Background(), tx)
	span, ctx := apm.StartSpan(ctx, name, "type")
	span.Duration = time.Second
	span.End()
	tx.Duration = 2 * time.Second
	tx.End()
	apm.DefaultTracer.Flush(nil)
	fmt.Printf("%s: %+v\n", name, apm.DefaultTracer.Stats())
	<-time.After(time.Minute)
}
```
7. Verify that the [configuration has been applied by the agent](http://localhost:5601/app/apm/settings/agent-configuration)
8. Navigate to Dev tools, issue the following query and verify that there's at least 1 document:
```javascript
GET metrics-apm.internal*/_search
{
  "query": {
    "term": {
      "metricset.name": {
        "value": "agent_config"
      }
    }
  }
}
```

## Related issues

Closes #6762
